internal/k8s/k8s.go

```diff
@@ -157,30 +164,74 @@ func New(cfg *Config) (*Client, error) {
 		c.syncFuncs = append(c.syncFuncs, c.svcInformer.HasSynced)
 
 		if cfg.ReadEndpoints {
-			epHandlers := cache.ResourceEventHandlerFuncs{
-				AddFunc: func(obj interface{}) {
-					key, err := cache.MetaNamespaceKeyFunc(obj)
-					if err == nil {
-						c.queue.Add(svcKey(key))
-					}
-				},
-				UpdateFunc: func(old interface{}, new interface{}) {
-					key, err := cache.MetaNamespaceKeyFunc(new)
-					if err == nil {
-						c.queue.Add(svcKey(key))
-					}
-				},
-				DeleteFunc: func(obj interface{}) {
-					key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-					if err == nil {
-						c.queue.Add(svcKey(key))
-					}
-				},
+			if !UseEndpointSlices(c.client) {
+				epHandlers := cache.ResourceEventHandlerFuncs{
+					AddFunc: func(obj interface{}) {
+						key, err := cache.MetaNamespaceKeyFunc(obj)
+						if err == nil {
+							c.queue.Add(svcKey(key))
+						}
+					},
+					UpdateFunc: func(old interface{}, new interface{}) {
+						key, err := cache.MetaNamespaceKeyFunc(new)
+						if err == nil {
+							c.queue.Add(svcKey(key))
+						}
+					},
+					DeleteFunc: func(obj interface{}) {
+						key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+						if err == nil {
+							c.queue.Add(svcKey(key))
+						}
+					},
+				}
+				epWatcher := cache.NewListWatchFromClient(c.client.CoreV1().RESTClient(), "endpoints", v1.NamespaceAll, fields.Everything())
+				c.epIndexer, c.epInformer = cache.NewIndexerInformer(epWatcher, &v1.Endpoints{}, 0, epHandlers, cache.Indexers{})
+
+				c.syncFuncs = append(c.syncFuncs, c.epInformer.HasSynced)
+			} else {
+				c.logger.Log("op", "New", "msg", "using endpoint slices")
+				slicesHandlers := cache.ResourceEventHandlerFuncs{
+					AddFunc: func(obj interface{}) {
+						slice, ok := obj.(*discovery.EndpointSlice)
+						if !ok {
+							return
```

Comment:
Should we be logging these errors?

Reply:
This is really unlikely to happen (or should always happen if we wire the controller in the wrong way). This is probably the reason why `epHandlers` didn't have it (and I followed the same logic), but I guess extra logging won't hurt.

Reply:
The only reason I thought so was that I couldn't see an easy way to detect if that failure happened.